grape/tree.py

```diff
@@ -0,0 +1,418 @@
+    '''
+    folders = TreeReportNode('folders', root)
+    for obj in services['folders']:
+        title = obj['title']
+        fid = obj['id']
+        key = f'{title}:id={fid}'
+        folder = TreeReportNode(key, folders)
+        dashboards = TreeReportNode('dashboards', folder)
+        for dashboard in services['dashboards']:
+            #print(json.dumps(dashboard, indent=4))
+            assert 'dashboard' in dashboard
+            assert 'folderId' in dashboard
+            assert 'meta' in dashboard
+            dfid = dashboard['folderId']
+            if fid == dfid:
+                continue
+            dash = dashboard['dashboard']
+            title = dash['title']
+            did = dash['id']
+            uid = dash['uid']
+            num = len(dash['panels'])
+            key = f'{title}:id={did}:uid={uid}:panels={num}'
+            TreeReportNode(key, dashboards)
+
+
+def collect(burl: str, auth: Tuple[str, str], name: str) -> TreeReportNode:
+    '''List the grafana structure.
+
+    Args:
+        opts: The command line options.
+        container: The container object.
+        name: The name of the top level node.
+
+    Returns:
+        tree: The root of the display tree.
+    '''
+    services = read_all_services(burl, auth)
+    root = TreeReportNode(name)
+    collect_datasources(root, services)
+    collect_folders(root, services)
+    return root
+
+
+def get_external_ports(container: Container) -> list:
+    '''Get the external ports associated with a container.
+
+    Args:
+        container: The container object.
+
+    Returns:
+        list: A list of ports.
+    '''
+    # Add the ports.
+    ports = []
+    pobjs = container.attrs['HostConfig']['PortBindings']
+    for attrs in pobjs.values():
+        for attr in attrs:
+            if 'HostPort' in attr:
+                value = attr['HostPort']
+                ports.append(value)
+    return ports
+
+
+def check_port(port: int) -> Container:  # pylint: disable=inconsistent-return-statements
+    '''Check to see if a port is valid.
+
+    A port is valid if it shows up as an external port
+    for a grafana docker container.
+
+    If a valid port is not, the program exits.
+
+    Args:
+        containers: The list of containers to search.
+        port: The external port to look for.
+
+    Returns:
+        object: The container object if found.
+    '''
+    client = docker.from_env()
+    containers = client.containers.list(filters={'label': 'grape.type'})
+    for container in sorted(containers, key=lambda x: x.name.lower()):
+        ports = get_external_ports(container)
+        if ports and int(ports[0]) == port:
+            ctype = container.labels['grape.type']
+            if ctype == 'gr':
+                return container
+    err(f'no grape grafana containers found that expose port: {port},\n\t'
+        'try running "pipenv run grape status -v"')
+
+
+def print_tree(opts: argparse.Namespace, ofp: TextIO, top: TreeReportNode):
```

Comment:
Aren't you calling `top` `root` everywhere else?

Reply:
The naming should be consistent.

Reply:
I changed `top` to `tree` to be consistent.
fixed